x509util/certificate.go

```diff
@@ -83,6 +83,58 @@ func NewCertificate(cr *x509.CertificateRequest, opts ...Option) (*Certificate,
 	return &cert, nil
 }
 
+// NewCertificateFromX509 creates a new Certificate from an x509.Certificate and
+// will apply template options. A new (unsigned) x509.CertificateRequest is created,
+// with date from the x509.Certificate template. This function is primarily useful
+// when signing a certificate for a key that can't sign a CSR or when the private
+// key is not available.
+func NewCertificateFromX509(template *x509.Certificate, opts ...Option) (*Certificate, error) {
+	// Copy data from the template to a new, unsigned CSR.
+	o, err := new(Options).apply(&x509.CertificateRequest{
+		PublicKey:          template.PublicKey,
+		PublicKeyAlgorithm: template.PublicKeyAlgorithm,
+		Subject:            template.Subject,
+		DNSNames:           template.DNSNames,
+		EmailAddresses:     template.EmailAddresses,
+		IPAddresses:        template.IPAddresses,
+		URIs:               template.URIs,
+		Extensions:         template.ExtraExtensions,
+	}, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// If no template use only the certificate request with the
+	// default leaf key usages.
+	if o.CertBuffer == nil {
+		return nil, errors.New("not implemented yet; use FromX509WithTemplate option")
+	}
+
+	// With templates
+	var cert Certificate
+	if err := json.NewDecoder(o.CertBuffer).Decode(&cert); err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling certificate")
+	}
+
+	// Enforce the public key from the template
+	cert.PublicKey = template.PublicKey
+	cert.PublicKeyAlgorithm = template.PublicKeyAlgorithm
+
+	// Generate the subjectAltName extension if the certificate contains SANs
+	// that are not supported in the Go standard library.
+	if cert.hasExtendedSANs() && !cert.hasExtension(oidExtensionSubjectAltName) {
+		ext, err := createCertificateSubjectAltNameExtension(cert, cert.Subject.IsEmpty())
+		if err != nil {
+			return nil, err
+		}
+		// Prepend extension to achieve a certificate as similar as possible to
+		// the one generated by the Go standard library.
+		cert.Extensions = append([]Extension{ext}, cert.Extensions...)
+	}
+
+	return &cert, nil
```

Comment:
This is duplicated logic that can be combined into one internal function.

Reply:
Fixed in f1025e376b1005740f087c1103cf7605b20eb175